src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributes.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    public class ReadOnlyTagHelperAttributes<TAttributeType> : IReadOnlyList<TAttributeType>
+        where TAttributeType : IReadOnlyTagHelperAttribute
+    {
+        protected readonly List<TAttributeType> _attributes;
+
+        protected ReadOnlyTagHelperAttributes()
+        {
+            _attributes = new List<TAttributeType>();
+        }
+
+        public ReadOnlyTagHelperAttributes([NotNull] IEnumerable<TAttributeType> attributes)
+        {
+            _attributes = new List<TAttributeType>(attributes);
+        }
+
+        public TAttributeType this[int index]
+        {
+            get
+            {
+                return _attributes[index];
+            }
+        }
+
+        public TAttributeType this[[NotNull] string key]
+        {
+            get
+            {
+                return _attributes.FirstOrDefault(attribute => KeyEquals(key, attribute));
```

Comment:
Was planning to optimize it, more concerned about the public API in this PR and if we're satisfied with what's exposed for the `TagHelper` author.